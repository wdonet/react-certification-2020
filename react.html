<html>
  <head>
    <title>MarioTube</title>
  </head>
  <body>
    <!--
        <div class="header">
            <div class="header-logo"><img src="./m-vector_icon.jpg" /></div>
            <div class="header-search">
                <label>Search videos <input type="text" name="search" placeholder="ðŸŽ¥  Search for videos"/></label>
            </div>
            <div class="header-login">
                <a>Sign in</a> <img src="./wize-logo.png" />
            </div>
        </div>
        <div class="sidebar">
          <ul>
            <li>Home</li>
            <li>Favorites</li>
          </ul>
          <div>Dark Mode</div>
        </div>
        <div class="content">
          <div class="categories">Categories</div>
          <div class="videos"><div class="video"></div></div>
        </div>
        <div class="footer"><div class="info"></div></div>

/**
* Sample JavaScript code for youtube.search.list
* See instructions for running APIs Explorer code samples locally:
* https://developers.google.com/explorer-help/guides/code_samples#javascript
*/-->
    <script src="https://apis.google.com/js/api.js"></script>
    <script>
      function authenticate() {
        console.log('authenticate...');
        return gapi.auth2
          .getAuthInstance()
          .signIn({ scope: 'https://www.googleapis.com/auth/youtube.force-ssl' })
          .then(
            function () {
              console.log('Sign-in successful');
            },
            function (err) {
              console.error('Error signing in', err);
            }
          );
      }
      function loadClient() {
        console.log('loadClient...');
        console.log('--setApiKey...');
        gapi.client.setApiKey('AIzaSyDpytkLvHqG_bnDtwfSklDRk-eU8O-fHL8');
        return gapi.client
          .load('https://www.googleapis.com/discovery/v1/apis/youtube/v3/rest')
          .then(
            function () {
              console.log('GAPI client loaded for API');
            },
            function (err) {
              console.error('Error loading GAPI client for API', err);
            }
          );
      }
      // Make sure the client is loaded and sign-in is complete before calling this method.
      function execute() {
        console.log('execute...');
        console.log('--search.list...');
        return gapi.client.youtube.search
          .list({
            part: ['snippet'],
            q: 'lacuna coil',
          })
          .then(
            function (response) {
              // Handle the results here (response.result has the parsed body).
              console.log('--Response', response);
            },
            function (err) {
              console.error('--Execute error', err);
            }
          );
      }
      gapi.load('client:auth2', function () {
        console.log('load...');
        console.log('--client:auth2...');
        console.log('--init...');
        gapi.auth2.init({
          client_id:
            '1088412977008-bt87ku9si9u48echferpm6i23rcv5jei.apps.googleusercontent.com',
        });
      });
    </script>
    <button onclick="authenticate().then(loadClient)">authorize and load</button>
    <button onclick="execute()">execute</button>

    <!--

    html, body, #root {
        width: 100%;
        height: 100%;
      }
      
      .SplitPane {
        width: 100%;
        height: 100%;
      }
      
      .SplitPane-left {
        float: left;
        width: 10%;
        height: 100%;
      }
      
      .SplitPane-right {
        float: left;
        width: 89%;
        height: 100%;
      }
      
      .Contacts {
        width: 100%;
        height: 100%; 
        background: lightblue;
      }
      
      .Chat {
        width: 100%;
        height: 100%; 
        background: pink;
      }


      function Contacts() {
        return <div className="Contacts" />;
      }
      
      function Chat() {
        return <div className="Chat" />;
      }
      
      function SplitPane(props) {
        return (
          <div className="SplitPane">
            <div className="SplitPane-left">
              {props.left}
            </div>
            <div className="SplitPane-right">
              {props.right}
            </div>
          </div>
        );
      }
      
      function App() {
        return (
          <SplitPane
            left={
              <Contacts />
            }
            right={
              <Chat />
            } />
        );
      }
      
      ReactDOM.render(
        <App />,
        document.getElementById('root')
      );
      

      YouTube App


      Menu Bar
          Icon    
          Search bar
          Login icon
              Login
              Account info
              Logout
      
      Side Bar
          Home
          Favorites
          Dark mode
      
      Main Body
          Videos
              Thumbnail
              Title
              Description
              Author
      
      Video View
          Video
              Title               Views
              Author              Like/Dislike
              Description
      
      Footer
      
      KNEWZ-1745
      
      React.createElement(
          "div",
          {},
          React.createElement("h1", {}, props.title),
          React.createElement("h1", {}, props.subtitle),
      )


      class LoginControl extends React.Component {
        constructor(props) {
            super(props);
            this.handleLoginClick = this.handleLoginClick.bind(this);
            this.handleLogoutClick = this.handleLogoutClick.bind(this);
            this.state = {isLoggedIn: false};
        }
      
        handleLoginClick() {
            this.setState({isLoggedIn: true});
        }
      
        handleLogoutClick() {
            this.setState({isLoggedIn: false});
        }
      
        render() {
            const isLoggedIn = this.state.isLoggedIn;
            let button;
      
            if (isLoggedIn) {
                button = <LogoutButton onClick={this.handleLogoutClick} />;
            } else {
                button = <LoginButton onClick={this.handleLoginClick} />;
            }
      
            return (
                <div>
                    <Greeting isLoggedIn={isLoggedIn} />
                    {button}
                </div>
            );
        }
    }
      
    function UserGreeting(props) {
        return <h1>Welcome back!</h1>;
    }
      
    function GuestGreeting(props) {
        return <h1>Please sign up.</h1>;
    }
      
    function Greeting(props) {
        const isLoggedIn = props.isLoggedIn;
        if (isLoggedIn) {
            return <UserGreeting />;
        }
        return <GuestGreeting />;
    }
      
    function LoginButton(props) {
        return (
            <button onClick={props.onClick}>
                Login
            </button>
        );
    }
      
    function LogoutButton(props) {
        return (
            <button onClick={props.onClick}>
                Logout
            </button>
        );
    }
      
    ReactDOM.render(
        <LoginControl />,
        document.getElementById('root')
    );
      
-->
  </body>
</html>
