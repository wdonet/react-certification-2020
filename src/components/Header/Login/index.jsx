import React from 'react';
import wizelogo from '../../../img/wize-logo.png';

const Login = () => {
  const authenticated = true;

  return (
    <a className="login-link" href="_blank">
      <img src={wizelogo} alt="wizelogo" height={60} />
      {authenticated ? <div className="login-user">Mario</div> : <div>Sign in</div>}
    </a>
  );
};

export default Login;
