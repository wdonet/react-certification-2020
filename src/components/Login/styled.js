import styled from 'styled-components';

const LoginBackground = styled.div`
  display: none;
  position: fixed;
  z-index: 1;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  background-color: rgb(0, 0, 0);
  background-color: rgba(0, 0, 0, 0.7);

  display: ${({ open }) => (open ? 'block' : 'none')};
`;

const LoginBox = styled.div`
  background-color: #fefefe;
  margin: 15% auto;
  padding: 20px;
  border: 1px solid #888;
  border-radius: 5%;
  width: 300px;
  height: 350px;
  position: relative;
`;

const CloseLoginButton = styled.span`
  color: #aaa;
  position: absolute;
  top: 0px;
  right: 15px;
  font-size: 28px;
  font-weight: bold;
  &:hover,
  &:focus {
    color: black;
    text-decoration: none;
    cursor: pointer;
  }
`;

const ErrorText = styled.p`
  color: red;
  font-size: 14px;
`;

const InputText = styled.input`
  width: 100%;
  height: 30px;
  border: 1px solid ${({ theme }) => theme.font.secondary};
  margin: 10px 0;
`;

const FormStyled = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 20px 0;
`;

export { LoginBackground, LoginBox, CloseLoginButton, ErrorText, InputText, FormStyled };
