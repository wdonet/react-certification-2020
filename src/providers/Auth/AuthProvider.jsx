import React, { useState, useEffect, useCallback } from 'react';

import { AUTH_STORAGE_KEY } from '../../utils/constants';
import { loginUser } from '../../mocks/loginApi';
import { storage } from '../../utils/storage';

const AuthContext = React.createContext(null);

function AuthProvider({ authenticatedJest, children }) {
  const [authenticated, setAuthenticated] = useState(authenticatedJest);

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY);
    const isAuthenticated = Boolean(lastAuthState);

    setAuthenticated(isAuthenticated || authenticatedJest);
  }, []);

  const login = async ({ username, password }) => {
    try {
      const authorized = await loginUser({ username, password });
      if (authorized) {
        setAuthenticated(true);
        storage.set(AUTH_STORAGE_KEY, true);
      }
      return { success: true };
    } catch (error) {
      return { error };
    }
  };

  const logout = useCallback(() => {
    setAuthenticated(false);
    storage.set(AUTH_STORAGE_KEY, false);
  }, []);

  return (
    <AuthContext.Provider value={{ login, logout, authenticated }}>
      {children}
    </AuthContext.Provider>
  );
}

export { AuthContext };
export default AuthProvider;
