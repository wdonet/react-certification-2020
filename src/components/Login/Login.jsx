import React, { useContext, useState } from "react";
import styled from "styled-components";
import loginService from '../../services/loginService';
import AppContext from "../../providers/AppContext";
import TextField from "../../ui/TextField/TextField";
import Button from "../../ui/Button/Button";
import { useHistory } from "react-router";

const StyledLogin = styled.div`
    border-radius: 8px;
    padding: 4px;
    margin: 4px;
    height: min-content;
    width: 400px;
    background: ${ ({theme}) => theme.color.surface };
    color: ${ ({theme}) => theme.color.fontPrimary };
`

const StyledSection = styled.div`
    display: ${({display}) => display ? display : 'grid'};
    height: min-content;
    margin: 8px;
`;

const Login = (props) => {
    const { push } = useHistory();
    const { theme, setUserSession } = useContext(AppContext);
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [error, setError] = useState(null);
    const { onCancel } = props;

    const cleanUp = () => {
        document.getElementById("username").value = "";
        document.getElementById("password").value = "";
        setError("");
    }

    const loginAttempt = async () => {
        try{
            setError(null);
            const response = await loginService(username, password);
            cleanUp();
            setUserSession(response);
            push({ pathname: "/home" });
            onCancel && onCancel();
        }catch(error){
            cleanUp();
            setError(error.message);
        }
    }

    return (<StyledLogin role="form" theme={theme} data-testid="login-form">
                <StyledSection>Login!</StyledSection>
                <StyledSection>
                    <label htmlFor="username">User</label>
                    <TextField 
                        id="username" 
                        name="username"
                        data-testid="username-input"
                        onChange={(uname) => setUsername(uname)}
                    />
                </StyledSection>
                <StyledSection>
                    <label htmlFor="password">Password</label>
                    <TextField 
                        id="password" 
                        name="password"
                        data-testid="password-input"
                        type="password"
                        onChange={(pword) => setPassword(pword)}
                    />
                </StyledSection>
                <StyledSection>
                <span style={{color:'red'}}>{ error }</span>
                </StyledSection>
                <StyledSection display="flex">
                    <Button 
                        data-testid="login-button"
                        onClick={loginAttempt}
                        margin="4px"
                        height="30px"
                    >
                        Login
                    </Button>
                    <Button 
                        data-testid="cancel-button"
                        onClick={() => onCancel && onCancel()}
                        margin="4px"
                        height="30px"
                    >
                        Cancel
                    </Button>
                </StyledSection>
            </StyledLogin>);
}

export default Login;