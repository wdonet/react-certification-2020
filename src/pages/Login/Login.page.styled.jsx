import styled from 'styled-components';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  flex: 1;
  margin-top: 5%;
`;

const ErrorLabel = styled.p`
  font-size: 15px;
  color: red;
`;

export { Container, ErrorLabel };
