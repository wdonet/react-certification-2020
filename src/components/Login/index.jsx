import React, { useEffect, useRef, useState } from 'react';

import {
  LoginBackground,
  LoginBox,
  CloseLoginButton,
  ErrorText,
  InputText,
  FormStyled,
} from './styled';

import { useAuth } from '../../providers/Auth';

const Login = ({ open, closeModal }) => {
  const [loginError, setLoginError] = useState(null);
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const { login } = useAuth();
  const loginBoxRef = useRef(null);
  useEffect(() => {
    const handleClickOutside = (e) => {
      if (loginBoxRef.current && !loginBoxRef.current.contains(e.target)) {
        closeModal();
      }
    };
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      // Unbind the event listener
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [loginBoxRef]);

  const loginUser = async () => {
    const { success } = await login({ username, password });
    if (!success) {
      setLoginError('Invalid username or password');
    } else {
      setLoginError(null);
      closeModal();
    }
  };

  const handleUsernameChange = (e) => {
    const {
      target: { value },
    } = e;
    setUsername(value);
  };

  const hanglePasswordChange = (e) => {
    const {
      target: { value },
    } = e;
    setPassword(value);
  };

  return (
    <LoginBackground open={open}>
      <LoginBox ref={loginBoxRef}>
        <CloseLoginButton onClick={closeModal}>&times;</CloseLoginButton>
        <FormStyled>
          <InputText
            name="username"
            value={username}
            placeholder="Username"
            onChange={handleUsernameChange}
          />
          <InputText
            name="password"
            placeholder="Password"
            type="password"
            value={password}
            onChange={hanglePasswordChange}
          />
          {loginError && (
            <ErrorText
              css={`
                color: ${({ theme }) => theme.palette.alert.main};
                padding: 2px;
                position: absolute;
                top: calc(100% - 25px);
                font-size: 16px;
              `}
            >
              {loginError}
            </ErrorText>
          )}
          <button
            type="button"
            onClick={() => {
              loginUser({ username: 'fulano', password: '1234567' });
              return false;
            }}
          >
            Login
          </button>
        </FormStyled>
      </LoginBox>
    </LoginBackground>
  );
};

export default Login;
