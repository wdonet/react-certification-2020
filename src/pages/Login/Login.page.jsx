import React, { useState } from 'react';

import { useAuth } from '../../providers/Auth';
import { Container, ErrorLabel } from './Login.page.styled';

function LoginPage() {
  const { login } = useAuth();
  const [values, setValues] = useState({});
  const [error, setError] = useState();

  const authenticate = async (event) => {
    event.preventDefault();
    try {
      setError(undefined);
      await login(values);
    } catch (err) {
      setError(err);
    }
  };

  const handleInputChange = (event) => {
    const newValues = { ...values };
    newValues[event.target.id] = event.target.value;
    setValues(newValues);
  };

  return (
    <Container>
      <h1>Welcome back!</h1>
      <form onSubmit={authenticate} className="login-form">
        <div className="form-group">
          <label htmlFor="username">
            <strong>username </strong>
            <input
              required
              type="text"
              id="username"
              value={values.username || ''}
              onChange={handleInputChange}
            />
          </label>
        </div>
        <div className="form-group">
          <label htmlFor="password">
            <strong>password </strong>
            <input
              required
              type="password"
              id="password"
              value={values.password || ''}
              onChange={handleInputChange}
            />
          </label>
        </div>
        <button type="submit">login</button>
      </form>
      {error && <ErrorLabel>Username or password incorrect!</ErrorLabel>}
    </Container>
  );
}

export default LoginPage;
