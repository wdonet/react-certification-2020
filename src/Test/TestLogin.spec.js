import React from 'react';
import { render, screen } from '@testing-library/react';
import AuthProvider from '../providers/Auth/Auth.provider';
import Login from '../pages/Login';

describe('Login Components', () => {
  it('User name input is present', () => {
    render(
      <AuthProvider>
        {' '}
        <Login />
      </AuthProvider>
    );
    expect(screen.getAllByRole('textbox').length).toBe(1);
  });

  it('H1 tag is present', () => {
    render(
      <AuthProvider>
        <Login />
      </AuthProvider>
    );
    expect(screen.getByText('Welcome back!')).toBeTruthy();
  });

  it('should take a snapshot', () => {
    const { asFragment } = render(
      <AuthProvider>
        <Login />
      </AuthProvider>
    );
    expect(
      asFragment(
        <AuthProvider>
          <Login />
        </AuthProvider>
      )
    ).toMatchSnapshot();
  });
});
