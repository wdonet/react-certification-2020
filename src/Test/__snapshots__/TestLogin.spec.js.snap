// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Components should take a snapshot 1`] = `
<DocumentFragment>
  <section
    class="login"
  >
    <h1>
      Welcome back!
    </h1>
    <form
      class="login-form"
    >
      <div
        class="form-group"
      >
        <label
          for="username"
        >
          <strong>
            username 
          </strong>
          <input
            id="username"
            required=""
            type="text"
          />
        </label>
      </div>
      <div
        class="form-group"
      >
        <label
          for="password"
        >
          <strong>
            password 
          </strong>
          <input
            id="password"
            required=""
            type="password"
          />
        </label>
      </div>
      <button
        type="submit"
      >
        login
      </button>
    </form>
  </section>
</DocumentFragment>
`;
